relentless/optimize/objective.py

```diff
@@ -1,84 +1,32 @@
-__all__ = ['RelativeEntropy']
+import abc
 
-import numpy as np
-import scipy.integrate
+from relentless import _collections
 
-from .core import OptimizationProblem
-from relentless._math import Interpolator
-from relentless.ensemble import Ensemble
+class ObjectiveFunction(abc.ABC):
 
-class RelativeEntropy(OptimizationProblem):
-    def __init__(self, engine, target, dr=0.01):
-        super().__init__(engine)
-        self.target = target
-        self.dr = dr
+    class Result:
+        def __init__(self, value, gradient, objective):
+            self.value = value
 
-    def _get_step(self, env, step):
-        # try loading the simulation
-        try:
-            with env.data(step):
-                thermo = Ensemble.load('ensemble')
-        except FileNotFoundError:
-            thermo = None
+            # gradient must be keyed only on objective design variables
+            self.__gradient = _collections.FixedKeyDict(keys=objective.design_variables())
+            self.__gradient.update(gradient)
+            self._gradient = self.__gradient.todict()
 
-        # if not found, run the simulation
-        if thermo is None:
-            # write the current parameters
-            with env.data(step):
-                for pot in self.engine.potentials:
-                    pot.save()
+        def gradient(self, var):
+            try:
+                return self._gradient[var]
+            except KeyError:
+                return 0.0
 
-            self.engine.run(env, step)
-            thermo = self.engine.process(env, step)
-            with env.data(step):
-                thermo.save('ensemble')
+        @property
+        def variables(self):
```

Comment:
For consistency, should this be:
```suggestion
        def design_variables(self):
```
?